### PR TITLE
Sketch fix of the performance issue of IDEEP

### PR DESCRIPTION
**Pull request by yinghai**
This is a quick fix of performance issues of IDEEP conv ops (https://github.com/intel/ideep/issues/7) . Just to show the idea. With this, I was able to get almost identical performance numbers with small batch size, compared to MKLML. And when the batch size is large (128), IDEEP numbers on resnet50 actually beats MKLML ops: 62 vs 58. Experimental was done with 1 socketful of 12 cores on a Broadwell CPU. 

It needs some clean up to make an official PR and note that I need to do the similar trick to fully-connected layer as well, although performance impact is not large.

### CODE REVIEW

File: caffe2/ideep/operators/conv_op.cc [hunk 1]
```diff
@@ -34,23 +36,43 @@ class IDEEPConvOp final : public IDEEPConvPoolOpBase {
...
+      cached_weights_descriptor_ = filter.get_descriptor();
+      filter_ = filter;
+      auto expected_descriptor =
+          ideep::convolution_forward::expected_weights_descriptor(
+              filter.get_dims());
+      if (filter_.get_descriptor() != expected_descriptor) {
+        filter_.init<ideep::utils::allocator, ideep::convolution_forward>(
+            expected_descriptor);
+        ideep::reorder::compute(filter, filter_);
+      }
```
Comment (yinghai) [inline caffe2/ideep/operators/conv_op.cc:51]: Key chunk of fix.

**Comment by yinghai** [comment]
Performance numbers (FPS) on resnet50 with 12 core 2.5GHz Broadwell socket. The larger the better. 

Device|1-thread/batch-128 | 12-thread/batch-128|1-thread/batch-1 | 12-thread/batch-1
-------|---------|----------|---------|----------
MKLML|7.93|58.42|7.40|49.72
IDEEP|8.27|62.76|7.93|49.65

**Comment by jgong5** [comment]
The result looks cool. But the solution would not work for training, right? In fact, I like your idea of updating the weights in the model opt path. That looks a workable solution for both training and inference. What do you think?

**Comment by yinghai** [comment]
I use `constant_weight` as a bypass for training.

**Comment by CaoZhongZ** [comment]
Yes, training. That requires a thought.

**Comment by CaoZhongZ** [comment]
I mean, we can consider SGD on weights with MKL-DNN's format, since element wise operations are agnostic to format. That requires a little more thought though.

**Comment by yinghai** [comment]
Right now, by adding `constant_weight=0` to the op's argument, we can revert to the old behavior and training should work. Well, this is a sketch fix, which doesn't require changes in IDEEP. :) I agree more thoughts are needed to have an elegant solution.

**Comment by jgong5** [comment]
@4pao yes, but since SGD is basically element-wise so as long as we make sure the formats of weights and gradients are the same, we may just fall back to the original CPU implementation. Of course, we may provide an optimized SGD op later.

@yinghai Does Caffe2 have a backend optimization path that does not require explicit calling of an extra   user-level API? I'm considering if we can make this weight replacement process transparent since it is applicable to both training and inference.

**Comment by yinghai** [comment]
> Does Caffe2 have a backend optimization path that does not require explicit calling of an extra user-level API? I'm considering if we can make this weight replacement process transparent since it is applicable to both training and inference.

Currently, we don't have that. We are considering adding such optimization in inference. If you'll contribute to the tensor rewriting part. I'll be glad to help review. It's not very difficult actually.

**Comment by jgong5** [comment]
@yinghai Sure, will do that.

**Comment by yinghai** [comment]
Closing this in favor of https://github.com/pytorch/pytorch/pull/7503.
